Add mortar development to contact action

recuero opened the pull request:
## Reason
Make recent implementations on mortar mechanical contact easier to use.

## Design
In principle, I propose keeping the existing mortar settings in the action under the `mortar_approach = legacy` option. The default approach, however, will be that based on mortar quantities such as normal gap and tangential velocities `mortar_approach = weighted`, which enter contact constraints using dual bases.

## Impact
Quicker creation of input files for advanced simulations. Lower entry point for analysts and new users.


This PR intends to pick up the work from #17511. 

Refs #16961, #17495

lindsayad committed:
Make weighted gap classes default mortar formulation classes

This is with respect to normal mechanical contact. Also make the default
value for `c_normal` something more likely to be appropriate for
realistic mechanics simulations

recuero committed:
Update contact mechanics action to include variationally consistent contact constraints with weighted quantities using dual bases.

The action allows for having two cases: "weighted", which uses weighted gaps and tangential velocities and dual bases; and
"legacy", which uses a combination of nodal and mortar constraint enforcement via 'NormalNodalLM' and 'TangentialMortarLM' objects,
mixing the interpolation orders of standard bases.

The performance of both approaches is very similar, but a small deterioration of the Jacobian when using the default 'weighted' approach
is possible. However, to get the full benefits of the mortar approach, selecting the default 'weighted' approach is strongly recommended.

Refs #16961, #17495

recuero committed:
Add frictional cylinder using contact mechanics action

recuero committed:
Update existing tests in varios ways:

 - Those that don't diff at all, make them 'weighted'
 - Those that diff, make them 'legacy'. Diffs usually come from: a) the 'weighted' option taking a few more iterations or
b) the 'weighted' option providing a more consistent result, for example, in terms of total number of nodes into contact
 - Add a couple of new tests that use friction through the mortar action (default mortar)

Some of the tests in the 'mortar_tm' folder should be converted to frictional pdass in the near future to hold key behavior.

Remove initial dual mortar test. Initial dual mortar test appears it was not making full use of dual bases. Application of dual bases
to mortar mechanical contact is now covered by other tests.

dschwen commented:
🐟

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/751127/) on 7b9dd3b wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/17980/site)

This comment will be updated on new commits.

dewenyushu reviewed (commented):
Looks good to me. In addition to the mortar approach options that we talked about, I just have several minor suggestions about using enum class.

recuero committed:
Address Dewen's comments.

Bring back _use_dual and switch moose enum.

Also, deprecate mortar 'legacy' option, to be removed in the future. Existing tests that use the legacy
option will be migrated to use the default mortar approach.

dewenyushu commented on modules/contact/src/actions/ContactAction.C:
```diff
@@ -108,8 +108,16 @@ ContactAction::validParams()
       "normal_lm_scaling", 1., "Scaling factor to apply to the normal LM variable");
   params.addParam<Real>(
       "tangential_lm_scaling", 1., "Scaling factor to apply to the tangential LM variable");
-  params.addParam<bool>("use_dual", false, "Whether to use the dual mortar approach");
```
Great! Thanks for keeping the option here!

recuero replied:
The addition of standard bases (`use_dual = false`) with the weighed approach assumes same-order bases, which is just one way to do it, since all combinations have shown some result instabilities (including mixed-order bases).

I'd envision that once/if we remove the `legacy` option from the action, we may not have to stress whether we are using dual bases or weighted-quantity constraints, so it may be less confusing then.

recuero committed:
Fix remaining tests

dschwen commented on modules/contact/src/actions/ContactAction.C:
```diff
@@ -108,9 +108,22 @@ ContactAction::validParams()
+      true,
+      "Whether to interpolate the nodal normals (e.g. classic idea of evaluating field at "
+      "quadrature points). If this is set to false, then non-interpolated nodal normals will be "
+      "used, and then the _normals member should be indexed with _i instead of _qp");
```
I think it's advisable to point out the `_i` vs. `_qp` part is aimed at code developers.

recuero replied:
yeah, this is something I would like to look into a bit more.

dschwen commented on modules/contact/src/actions/ContactAction.C:
```diff
@@ -368,72 +395,161 @@ ContactAction::addMortarContact()
     };
 
-    // Normal contact: same family/order as primal.
+    // To avoid instabilities in the solution and obtain the full benefits of a variational
+    // enforcement, use of dual mortar is strongly recommended.
```
Any reason not to deprecate the `use_dual` parameter and _always_ use dual bases?

recuero replied:
Some developers may be still making comparisons between dual and standard bases (even if it's just for research purposes). But, from a user point of view, we should do that soon to avoid misuse (note that now, to use mortar, a user should only do `formulation = mortar`). In this code I am deprecating the previous way of doing mortar contact, so we can remove `_use_dual` and the `legacy` approach a few months from now.

recuero replied:
https://github.com/idaholab/moose/pull/17980#discussion_r643388815

dschwen commented on modules/contact/test/tests/bouncing-block-contact/frictional-nodal-min-normal-lm-mortar-pdass-tangential-lm-mortar-action.i:
```diff
@@ -0,0 +1,149 @@
+    type = CumulativeValuePostprocessor
+    postprocessor = num_nl
+  [../]
+ [contact]
```
It would be nice to have new input files `hit format`ed

dschwen commented on modules/contact/src/actions/ContactAction.C:
```diff
@@ -95,9 +95,9 @@ ContactAction::validParams()
                         "The tolerance of the frictional force for augmented Lagrangian method.");
   params.addParam<Real>(
-      "c_normal", 1, "Parameter for balancing the size of the gap and contact pressure");
+      "c_normal", 1e6, "Parameter for balancing the size of the gap and contact pressure");
```
Any helpful comments to guid the user's choice of that parameter? I'm thinking this is probably unit related?

recuero replied:
Added more explanation. But, for now, we may need to tune it for each type of problem.

recuero replied:
Yes, it is unit related. It has to do with the updates to the contact constraints.

dschwen commented on modules/contact/src/actions/ContactAction.C:
```diff
@@ -336,19 +357,25 @@ ContactAction::addMortarContact()
+      const int lm_order = std::max(primal_type.order.get_order() - codimension, min_lm_order);
 
-      if (primal_type.family == MONOMIAL || (primal_type.family == LAGRANGE && lm_order < 1))
+      if ((primal_type.family == LAGRANGE) && (lm_order < 1))
```
```suggestion
      if (primal_type.family == LAGRANGE && lm_order < 1)
```

if you're changing stuff anyways

dschwen commented on modules/contact/src/actions/ContactAction.C:
```diff
@@ -368,72 +395,161 @@ ContactAction::addMortarContact()
+    //   order of 1 (we don't have zeroth order Lagrange). We must use a Lagrange basis because we
+    //   need dofs at nodes in order to enforce the zero-penetration constraint
+    //
+    if (_mortar_approach == MortarApproach::Weighted)
+      addLagrangeMultiplier(
+          normal_lagrange_multiplier_name, 1, false, getParam<Real>("normal_lm_scaling"));
+    else if (_mortar_approach == MortarApproach::Legacy)
+      addLagrangeMultiplier(
+          normal_lagrange_multiplier_name, 1, false, getParam<Real>("normal_lm_scaling"));
```
this seems to be doing the same thing in both branches

```suggestion
    if (_mortar_approach == MortarApproach::Weighted || 
        _mortar_approach == MortarApproach::Legacy)
      addLagrangeMultiplier(
          normal_lagrange_multiplier_name, 1, false, getParam<Real>("normal_lm_scaling"));
```

recuero replied:
Probably we can just fully remove the `if` statement here since it is enclosed in the `mortar` formulation.

dschwen commented on modules/contact/src/actions/ContactAction.C:
```diff
@@ -368,72 +395,161 @@ ContactAction::addMortarContact()
+        // secondary_disp_z is not implemented for tangential (yet).
+
+        _problem->addConstraint(
+            "ComputeFrictionalForceLMMechanicalContact", action_name + "_tangential_lm", params);
```
This could probably be simplified as there is quite a bit of shared code between the coulomb/!coulomb branches (i.e. swap the nesting order of the conditionals)

dschwen replied:
never mind, not worth the hassle

recuero commented on modules/contact/test/tests/non-singular-frictional-mortar/tests:
```diff
@@ -3,11 +3,12 @@
-    expect_out = '0 of 41 singular values'
-    absent_out = '[1-9]+[0-9]* of 41 singular values'
-    mesh_mode = REPLICATED
+    design = 'ComputeFrictionalForceLMMechanicalContact.md'
+    issues = '#17495'
```
Updating the issue number is pertinent here, isn't it?

recuero committed:
Address Daniel's review.

Apply hit format to touched/created input files. Simplify contact action logic, add comments on numerical parameter.

recuero committed:
Remove unnecessary parentheses from contact action logic

Co-authored-by: Daniel Schwen <daniel@schwen.de>

lindsayad commented on modules/contact/src/actions/ContactAction.C:
```diff
@@ -336,19 +363,25 @@ ContactAction::addMortarContact()
+
+      // Allow the user to select "weighted" constraints and standard bases (use_dual = false)
+      // Unless it's for testing purposes, this combination isn't recommended
+      params.set<bool>("use_dual") = (_mortar_approach == MortarApproach::Weighted && _use_dual);
```
So if a `ContactAction` user specifies `use_dual` in their input but they're not using `MortarApproach::Weighted`, then their choice is not honored? Is this a change in behavior?

recuero replied:
It is a change of behavior that isn't present in tests apparently. I made changes to avoid this change of behavior.

lindsayad commented on modules/contact/src/actions/ContactAction.C:
```diff
@@ -368,72 +401,157 @@ ContactAction::addMortarContact()
     };
 
-    // Normal contact: same family/order as primal.
+    // To avoid instabilities in the solution and obtain the full benefits of a variational
+    // enforcement, use of dual mortar is strongly recommended.
```
This comment about dual mortar seems like it's kind of floating in space, not really close to any lines of code that do anything with dual

recuero replied:
I intended to put it before we create the Lagrange's multipliers, where you have other relevant comments. I moved it to the description of the `_use_dual` parameters, which might make sense while we have it.

lindsayad commented on modules/contact/src/actions/ContactAction.C:
```diff
@@ -462,7 +580,7 @@ ContactAction::addMortarContact()
     };
 
-    // Add mortar mechanical contact for each dimension
+    // Add residual contribution from mortar mechanical contact on primal variables
```
```suggestion
    // Add mortar mechanical contact constraint objects for primal variables
```

lindsayad reviewed (changes requested):
Very minor stuff. Overall looks good to me!

recuero committed:
Correct comment on constraint object

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

recuero committed:
Address Alex's review
